diff --git a/scripts/txt2img.py b/scripts/txt2img.py
index ef52ee0..d7791b6 100644
--- a/scripts/txt2img.py
+++ b/scripts/txt2img.py
@@ -263,6 +263,8 @@
                                 base_count += 1
                         all_samples.append(x_samples_ddim)
 
+                        import habana_frameworks.torch.core as htcore
+                        htcore.mark_step()
                 if not opt.skip_grid:
                     # additionally, save as grid
                     grid = torch.stack(all_samples, 0)
